### PR TITLE
Heedls 607 course delegates export all

### PR DESCRIPTION
**Pull request by AlexJacksonDS**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-607

### Description
Added the Export All excel download to the Delegate Courses page. This outputs an excel file for all delegates on the currently displayed courses.

Note that this should make use of the new search/sort/filter service so, depending on what gets merged first, that change needs to be made.

### Screenshots
image

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/Services/CourseDelegatesDownloadFileServiceTests.cs [hunk 1]
```diff
@@ -18,6 +19,9 @@ public class CourseDelegatesDownloadFileServiceTests
         public const string CourseDelegateExportCurrentDataDownloadRelativeFilePath =
             "\\TestData\\CourseDelegateExportCurrentDataDownloadTest.xlsx";
 
+        public const string CourseDelegateExportAllDataDownloadRelativeFilePath =
```
Comment (GeorgeBlackwell) [inline DigitalLearningSolutions.Data.Tests/Services/CourseDelegatesDownloadFileServiceTests.cs:22]: Do these need to be public?

File: DigitalLearningSolutions.Web/Scripts/trackingSystem/delegateCourses.ts [hunk 2]
```diff
@@ -1,4 +1,18 @@
...
+import { getQuery } from '../searchSortFilterAndPaginate/search';
+
+const exportAllLink = <HTMLAnchorElement>document.getElementById('export-all');
+exportAllLink.addEventListener('click', () => {
+  const searchString = getQuery();
+  const existingFilterString = getExistingFilterStringValue();
+  const sortBy = getSortBy();
+  const sortDirection = getSortDirection();
+  const pathWithCurrentSortFilter = getPathForEndpoint(`TrackingSystem/Delegates/Courses/DownloadAll?searchString=${searchString}&sortBy=${sortBy}&sortDirection=${sortDirection}&existingFilterString=${existingFilterString}`);
+  exportAllLink.href = pathWithCurrentSortFilter;
```
Comment (GeorgeBlackwell) [inline DigitalLearningSolutions.Web/Scripts/trackingSystem/delegateCourses.ts:14]: You could get away with assigning `exportAllLink.href` directly here

**Review by GeorgeBlackwell — CHANGES_REQUESTED** [review]
Nice work, just a few minor comments

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateCoursesController.cs [hunk 3]
```diff
@@ -78,5 +83,32 @@ public IActionResult AllCourseStatistics()
...
+            var content = courseDelegatesDownloadFileService.GetCourseDelegateDownloadFile(
+                centreId,
+                categoryId,
+                searchString,
+                sortBy,
+                existingFilterString,
+                sortDirection
+            );
+
+            const string fileName = "Digital Learning Solutions Course Delegates.xlsx";
```
Comment (JBsoftwire) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateCoursesController.cs:106]: Unless this filename has been specified I think it wants some kind of identifying detail, for example the datetime. Right now there'll be no way to distinguish different downloads.

**Review by JBsoftwire — CHANGES_REQUESTED** [review]
Looks good, but I think the filename might create problems for users.

**Review by GeorgeBlackwell — APPROVED** [review]
Nice work!

**Pull request merged**